cranelift-codegen/src/legalizer/boundary.rs

```diff
@@ -265,6 +265,152 @@ where
+            .filter(|r| r.purpose != ArgumentPurpose::Normal)
+            .collect();
+        let new_special_params: Vec<_> = pos.func.dfg.signatures[sig_ref]
+            .params
+            .iter()
+            .filter(|r| r.purpose != ArgumentPurpose::Normal)
+            .collect();
+        debug_assert_eq!(old_special_params.len() + 1, new_special_params.len());
+        debug_assert!(old_special_params
+            .iter()
+            .zip(&new_special_params)
+            .all(|(old, new)| old.purpose == new.purpose));
+        debug_assert_eq!(
+            new_special_params.last().unwrap().purpose,
+            ArgumentPurpose::StructReturn
+        );
+
+        let old_special_returns: Vec<_> = old_sig
+            .returns
+            .iter()
+            .filter(|r| r.purpose != ArgumentPurpose::Normal)
+            .collect();
+        let new_special_returns: Vec<_> = pos.func.dfg.signatures[sig_ref]
+            .returns
+            .iter()
+            .filter(|r| r.purpose != ArgumentPurpose::Normal)
+            .collect();
+        debug_assert!(old_special_returns
+            .iter()
+            .zip(&new_special_returns)
+            .all(|(old, new)| old.purpose == new.purpose));
+        debug_assert!(
+            old_special_returns.len() == new_special_returns.len()
+                || (old_special_returns.len() + 1 == new_special_returns.len()
+                    && new_special_returns.last().unwrap().purpose
+                        == ArgumentPurpose::StructReturn)
+        );
+    }
+
+    // Go through and remove all normal return values from the `call`
+    // instruction's returns list. These will be stored into the stack slot that
+    // the sret points to. At the same time, calculate the size and alignment
+    // required for the sret stack slot.
+    let mut sret_slot_size = 0;
+    let mut max_align = 0;
+    for (i, ret) in old_sig.returns.iter().enumerate() {
+        let v = old_ret_list.get(i, &pos.func.dfg.value_lists).unwrap();
+        let ty = pos.func.dfg.value_type(v);
+        if ret.purpose == ArgumentPurpose::Normal {
+            debug_assert_eq!(ret.location, ArgumentLoc::Unassigned);
+            let ty = legalized_type_for_sret(ty);
+            let size = ty.bytes();
+            let align = size;
+            max_align = std::cmp::max(max_align, align);
+            sret_slot_size = round_up_to_pow2(sret_slot_size, align) + size;
+        } else {
+            let new_v = pos.func.dfg.append_result(call, ty);
+            pos.func.dfg.change_to_alias(v, new_v);
+        }
+    }
+
+    // Now that we know the size and alignment required of the sret's stack
+    // slot, create it and insert the `sret` argument in the call.
+    let stack_offset = pos.func.stack_slots.values().fold(0, |off, slot| {
+        let slot_size = slot.size as i32;
+        let slot_offset = slot.offset.unwrap_or(-slot_size);
+        std::cmp::max(off, (slot_offset + slot_size) as u32)
```

Comment:
I found that they have been, and that if I set the offset to `None`, then things would break later on because they assumed it had already been set, but I didn't know whether I should assert that the offsets are already set here or not.

Reply:
I've thought about it: this function will be called during legalization, so well before register allocation. The latter might introduce spill slots, and since the ordering (as described in `stack_layout.rs`) puts the explicit slots / spill slots *before* the incoming return stack space, I wonder if a spill after a call to a multi-return function would cause an incorrect offset here (leading to incorrect behavior, probably stack clobbering)? If so, I imagine the fix would be to put the struct stack space *before* the explicit / spill slots, and ignore stack slots of kinds appearing after the stack-ret-space in this offset computation.

Reply:
Ok, I've switched things up a little and removed a bunch of code, and now struct return slots are not given pre-assigned offsets in the frame, instead they are assigned offsets by the existing frame layout algorithm. Nice little clean up, thanks for pushing back on this!